engine/execution/computation/metrics/transaction_execution_metrics.go

```diff
@@ -0,0 +1,168 @@
+	// all non-finalized metrics for that height are discarded
+	*collector
+
+	// provider is responsible for providing the metrics for the rpc endpoint
+	// it has a circular buffer of metrics for the last N finalized and executed blocks.
+	*provider
+
+	component.Component
+	// transactionExecutionMetricsProvider needs to consume BlockFinalized events.
+	psEvents.Noop
+
+	log zerolog.Logger
+
+	executionState         state.FinalizedExecutionState
+	headers                storage.Headers
+	blockFinalizedNotifier engine.Notifier
+
+	latestFinalizedAndExecuted *flow.Header
+}
+
+var _ TransactionExecutionMetricsProvider = (*transactionExecutionMetricsProvider)(nil)
+
+func NewTransactionExecutionMetricsProvider(
+	log zerolog.Logger,
+	executionState state.FinalizedExecutionState,
+	headers storage.Headers,
+	latestFinalizedBlock *flow.Header,
+	bufferSize uint,
+) TransactionExecutionMetricsProvider {
+	log = log.With().Str("component", "transaction_execution_metrics_provider").Logger()
+
+	collector := newCollector(log, latestFinalizedBlock.Height)
+	provider := newProvider(log, bufferSize, latestFinalizedBlock.Height)
+
+	p := &transactionExecutionMetricsProvider{
+		collector:                  collector,
+		provider:                   provider,
+		log:                        log,
+		executionState:             executionState,
+		headers:                    headers,
+		blockFinalizedNotifier:     engine.NewNotifier(),
+		latestFinalizedAndExecuted: latestFinalizedBlock,
+	}
+
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(collector.metricsCollectorWorker)
+	cm.AddWorker(p.blockFinalizedWorker)
+
+	p.Component = cm.Build()
+
+	return p
+}
+
+func (p *transactionExecutionMetricsProvider) BlockFinalized(*flow.Header) {
+	p.blockFinalizedNotifier.Notify()
+}
+
+// move data from the collector to the provider
+func (p *transactionExecutionMetricsProvider) onBlockExecutedAndFinalized(block flow.Identifier, height uint64) {
+	data := p.collector.Pop(height, block)
+	p.provider.Push(height, data)
+}
+
+func (p *transactionExecutionMetricsProvider) blockFinalizedWorker(
+	ctx irrecoverable.SignalerContext,
+	ready component.ReadyFunc,
+) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-p.blockFinalizedNotifier.Channel():
+			p.onExecutedAndFinalized()
+		}
+	}
+}
+
+func (p *transactionExecutionMetricsProvider) onExecutedAndFinalized() {
+	latestFinalizedAndExecutedHeight, err := p.executionState.GetHighestFinalizedExecuted()
+
+	if err != nil {
+		p.log.Warn().Err(err).Msg("could not get highest finalized executed")
+		return
+	}
+
+	// the latest finalized and executed block could be more than one block further than the last one handled
+	// step through all blocks between the last one handled and the latest finalized and executed
+	for height := p.latestFinalizedAndExecuted.Height + 1; height <= latestFinalizedAndExecutedHeight; height++ {
+		header, err := p.headers.ByHeight(height)
```

Comment:
```suggestion
		blockID, err := p.headers.BlockIDByHeight(height)
```

We don't need the header, but only the ID. And we also have the height.